### PR TITLE
[SPARK-29768][SQL] Column pruning through nondeterministic expressions 

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

Support columnar pruning through non-deterministic expressions.

### Why are the changes needed?

In some cases, columns can still be pruned even though nondeterministic expressions appears.
e.g. for the plan  `Filter('a = 1, Project(Seq('a, rand() as 'r), LogicalRelation('a, 'b)))`, we shall still prune column b while non-deterministic expression appears. 

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added a new test file: `ScanOperationSuite`.
Added test in `FileSourceStrategySuite` to verify the right prune behavior for both DS v1 and v2.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] {
   import DataSourceV2Implicits._
 
   override def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
-    case PhysicalOperation(project, filters, relation: DataSourceV2Relation) =>
+    case FileSourceOperation(project, filters, relation: DataSourceV2Relation) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala]: This is not just file scan. How about `ScanOperation`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 2]
```diff
@@ -74,24 +95,66 @@ object PhysicalOperation extends PredicateHelper {
...
+object FileSourceOperation extends OperationHelper with PredicateHelper {
 
-  private def substitute(aliases: AttributeMap[Expression])(expr: Expression): Expression = {
-    expr.transform {
-      case a @ Alias(ref: AttributeReference, name) =>
-        aliases.get(ref)
-          .map(Alias(_, name)(a.exprId, a.qualifier))
-          .getOrElse(a)
+  private val invalid: LogicalPlan => (Option[Seq[NamedExpression]], Seq[Expression],
+    LogicalPlan, Boolean, AttributeMap[Expression], Set[String]) =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: we can create an alias for this type

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 3]
```diff
@@ -74,24 +95,66 @@ object PhysicalOperation extends PredicateHelper {
...
-    })
+object FileSourceOperation extends OperationHelper with PredicateHelper {
 
-  private def substitute(aliases: AttributeMap[Expression])(expr: Expression): Expression = {
-    expr.transform {
-      case a @ Alias(ref: AttributeReference, name) =>
-        aliases.get(ref)
-          .map(Alias(_, name)(a.exprId, a.qualifier))
-          .getOrElse(a)
+  private val invalid: LogicalPlan => (Option[Seq[NamedExpression]], Seq[Expression],
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: another idea: maybe we can make the entire thing `Option`, and use `None` to indicate invalid.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 4]
```diff
@@ -74,24 +95,70 @@ object PhysicalOperation extends PredicateHelper {
...
-    })
-
-  private def substitute(aliases: AttributeMap[Expression])(expr: Expression): Expression = {
-    expr.transform {
-      case a @ Alias(ref: AttributeReference, name) =>
-        aliases.get(ref)
-          .map(Alias(_, name)(a.exprId, a.qualifier))
-          .getOrElse(a)
+/**
+ * Unlike PhysicalOperation, this operation would firstly search the first non-Project
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: A variant of [[PhysicalOperation]]. It matches any number of project or filter operations even if they are non-deterministic, as long as they object the requirement of `CollapseProject` and `CombineFilters`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 5]
```diff
@@ -74,24 +95,70 @@ object PhysicalOperation extends PredicateHelper {
...
+      case Alias(ref: AttributeReference, _) if aliases.contains(ref) =>
+        aliases(ref)
+      case a: AttributeReference if aliases.contains(a) =>
+        aliases(a)
+    }.exists(!_.deterministic))
+  }
+
+  private def collectProjectsAndFilters(plan: LogicalPlan)
+    : Option[(Option[Seq[NamedExpression]], Seq[Expression], LogicalPlan,
+    AttributeMap[Expression])] =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: it's hard to read when the type is too long to fit one line. Can we create a type alias?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 6]
```diff
@@ -74,24 +95,70 @@ object PhysicalOperation extends PredicateHelper {
...
+      case Some((fields, filters, child, _)) =>
+        Some((fields.getOrElse(child.output), filters, child))
+      case None => None
     }
   }
+
+  private def hasCommonNonDeterministic(expr: Seq[Expression], aliases: AttributeMap[Expression])
+    : Boolean = {
+    expr.exists(_.collect {
+      case Alias(ref: AttributeReference, _) if aliases.contains(ref) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: do we need to match Alias? `TreeNode.collect` will recursively check the children

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 7]
```diff
@@ -74,24 +95,70 @@ object PhysicalOperation extends PredicateHelper {
...
+        aliases(a)
+    }.exists(!_.deterministic))
+  }
+
+  private def collectProjectsAndFilters(plan: LogicalPlan): ScanReturnType =
+      plan match {
+        case Project(fields, child) =>
+          collectProjectsAndFilters(child) match {
+            case Some((_, filters, other, aliases)) =>
+              if (!hasCommonNonDeterministic(fields, aliases)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: let's add some comments to explain the logic. e.g. follow `CollapseProject` and only keep going if the collected `Project`s do not have common non-deterministic expressions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 8]
```diff
@@ -74,24 +95,70 @@ object PhysicalOperation extends PredicateHelper {
...
+              } else {
+                None
+              }
+            case None => None
+          }
+
+        case Filter(condition, child) =>
+          collectProjectsAndFilters(child) match {
+            case Some((fields, filters, other, aliases)) =>
+              if (filters.forall(_.deterministic) &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/planning/ScanOperationSuite.scala [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+case class TestRelation() extends LeafNode {
+  override def output: Seq[Attribute] = Seq(
+    AttributeReference("a", StringType)(),
+    AttributeReference("b", StringType)(),
+    AttributeReference("c", StringType)())
+}
+
+class ScanOperationSuite extends SparkFunSuite {
+  test("Collect projects and filters through non-deterministic expressions") {
+    OneRowRelation
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/planning/ScanOperationSuite.scala]: unnecessary line?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 10]
```diff
@@ -264,23 +264,23 @@ case class DataSourceStrategy(conf: SQLConf) extends Strategy with Logging with
   import DataSourceStrategy._
 
   def apply(plan: LogicalPlan): Seq[execution.SparkPlan] = plan match {
-    case PhysicalOperation(projects, filters, l @ LogicalRelation(t: CatalystScan, _, _, _)) =>
+    case ScanOperation(projects, filters, l @ LogicalRelation(t: CatalystScan, _, _, _)) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: Let's test ds v1 in `PrunedScanSuite`

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PrunedScanSuite.scala [hunk 11]
```diff
@@ -148,5 +148,27 @@ class PrunedScanSuite extends DataSourceTest with SharedSparkSession {
...
+
+  testNonDeterministicPruning("SELECT a, rand() FROM oneToTenPruned WHERE a > 5", "a")
+
+  testNonDeterministicPruning("SELECT a FROM oneToTenPruned WHERE rand() > 5", "a")
+
+  testNonDeterministicPruning("SELECT a, rand() FROM oneToTenPruned WHERE rand() > 5", "a")
+
+  testNonDeterministicPruning("SELECT a, rand() FROM oneToTenPruned WHERE b > 5", "a", "b")
+
+  def testNonDeterministicPruning(sqlString: String, expectedColumns: String*): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PrunedScanSuite.scala]: why can't we reuse `testPruning`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 12]
```diff
@@ -74,24 +95,72 @@ object PhysicalOperation extends PredicateHelper {
...
-  private def substitute(aliases: AttributeMap[Expression])(expr: Expression): Expression = {
-    expr.transform {
-      case a @ Alias(ref: AttributeReference, name) =>
-        aliases.get(ref)
-          .map(Alias(_, name)(a.exprId, a.qualifier))
-          .getOrElse(a)
+/**
+ * A variant of [[PhysicalOperation]]. It matches any number of project or filter
+ * operations even if they are non-deterministic, as long as they satisfy the
+ * requirement of CollapseProject and CombineFilters.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: To check this is satisfied is caller's responsibility?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 13]
```diff
@@ -74,24 +95,72 @@ object PhysicalOperation extends PredicateHelper {
...
+  def unapply(plan: LogicalPlan): Option[ReturnType] = {
+    collectProjectsAndFilters(plan) match {
+      case Some((fields, filters, child, _)) =>
+        Some((fields.getOrElse(child.output), filters, child))
+      case None => None
     }
   }
+
+  private def hasCommonNonDeterministic(expr: Seq[Expression], aliases: AttributeMap[Expression])
+    : Boolean = {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: you can split at the parameters, it looks better.

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Something like `hasCommonNonDeterministic(expr: Seq[Expression])(aliases: AttributeMap[Expression])` ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala (hunk 12 above)
Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: No, `ScanOperation.collectProjectsAndFilters()` does the check by itself.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala (hunk 13 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Like:
```scala
private def hasCommonNonDeterministic(
    expr: Seq[Expression],
    aliases: AttributeMap[Expression]): Boolean = {
  ...
}
```

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Ah, I see. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 14]
```diff
@@ -74,24 +95,72 @@ object PhysicalOperation extends PredicateHelper {
...
+    }.exists(!_.deterministic))
+  }
+
+  private def collectProjectsAndFilters(plan: LogicalPlan): ScanReturnType =
+      plan match {
+        case Project(fields, child) =>
+          collectProjectsAndFilters(child) match {
+            case Some((_, filters, other, aliases)) =>
+              // Follow CollapseProject and only keep going if the collected Projects
+              // do not have common non-deterministic expressions.
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: if any projects in the child has no common non-deterministic expressions, aren't they already combined by `CollapseProject` rule?

Comment (Ngone51) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: Yeah, generally, I think so. But there's also suggestion of reducing coupling between `Optimizer` and `Planner`. See: https://github.com/apache/spark/pull/8486#issuecomment-136518354

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala [hunk 15]
```diff
@@ -74,24 +95,73 @@ object PhysicalOperation extends PredicateHelper {
...
+  private def hasCommonNonDeterministic(
+      expr: Seq[Expression],
+      aliases: AttributeMap[Expression]): Boolean = {
+    expr.exists(_.collect {
+      case a: AttributeReference if aliases.contains(a) => aliases(a)
+    }.exists(!_.deterministic))
+  }
+
+  private def collectProjectsAndFilters(plan: LogicalPlan): ScanReturnType =
+      plan match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala]: nit: `def .... = plan match {` or
```
def .. = {
  plan match ...
}
```

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by Ngone51** [comment]
Thanks!! @cloud-fan @viirya

**Pull request closed without merging**